contracts/signature_mint/SignatureMint.sol

```diff
@@ -0,0 +1,276 @@
+
+    /// @dev The token ID of the next token to mint.
+    uint256 public nextTokenIdToMint;
+
+    /// @dev Contract interprets 10_000 as 100%.
+    uint64 private constant MAX_BPS = 10_000;
+
+    /// @dev The % of secondary sales collected as royalties. See EIP 2981.
+    uint64 public royaltyBps;
+
+    /// @dev The % of primary sales collected by the contract as fees.
+    uint120 public feeBps;
+
+    /// @dev Whether transfers on tokens are restricted.
+    bool public transfersRestricted;
+
+    /// @dev Contract level metadata.
+    string public contractURI;
+    
+    /// @dev The protocol control center.
+    ProtocolControl internal controlCenter;
+
+    uint256[] private baseURIIndices;
+    
+    /// @dev Mapping from end-tokenId => baseURI.
+    mapping(uint256 => string) public baseURI;
+
+    /// @dev Mapping from mint request UID => whether the mint request is processed.
+    mapping(bytes => bool) private minted;
+
+    modifier onlyModuleAdmin() {
+        require(hasRole(DEFAULT_ADMIN_ROLE, _msgSender()), "not module admin.");
+        _;
+    }
+    
+    constructor(
+        string memory _name,
+        string memory _symbol,
+        string memory _contractURI,
+        address payable _controlCenter,
+        address _trustedForwarder,
+        uint128 _royaltyBps,
+        uint128 _feeBps
+    ) 
+        ERC721(_name, _symbol) 
+        EIP712(_name, "1")
+        ERC2771Context(_trustedForwarder)
+    {
+        // Set the protocol control center
+        controlCenter = ProtocolControl(_controlCenter);
+        contractURI = _contractURI;
+        royaltyBps = uint64(_royaltyBps);
+        feeBps = uint120(_feeBps);
+
+        address deployer = _msgSender();
+        _setupRole(DEFAULT_ADMIN_ROLE, deployer);
+        _setupRole(MINTER_ROLE, deployer);
+    }
+
+    ///     =====   Public functions  =====
+
+    /// @dev Verifies that a mint request is signed by an account holding MINTER_ROLE (at the time of the function call).
+    function verify(
+        MintRequest calldata req,
+        bytes calldata signature
+    )
+        public
+        view 
+        returns (bool)
+    {
+        address signer = _hashTypedDataV4(
+            keccak256(abi.encode(TYPEHASH, req.to, req.amountToMint, req.validityStartTimestamp, req.validityEndTimestamp, keccak256(req.uid)))
+        ).recover(signature);
+
+        return !minted[req.uid] && hasRole(MINTER_ROLE, signer);
+    }
+
+    /// @dev Returns the URI for a given tokenId.
+    function tokenURI(uint256 _tokenId) public view override returns (string memory) {
+        for (uint256 i = 0; i < baseURIIndices.length; i += 1) {
+            if (_tokenId < baseURIIndices[i]) {
+                return string(abi.encodePacked(baseURI[baseURIIndices[i]], _tokenId.toString()));
+            }
+        }
+
+        return "";
+    }
+
+    ///     =====   External functions  =====
+
+    function mint(MintRequest calldata _req, bytes calldata _signature) external {
```

Comment:
I think we should allow creator to set payment. Either in the payload or in the contract :) pricePerToken + currency. I think it unlocks a lot of different capability in the product.
wyt?

Reply:
I'll accept `pricePerToken` and `currency` fields in the MintRequest struct.